active/0000-production-code-stripping.md

```diff
@@ -0,0 +1,56 @@
+- Start Date: 2016-04-06
+- RFC PR: (leave this empty)
+- ember-cli Issue: (leave this empty)
+
+# Summary
+
+A number of Ember framework function calls are no-ops in production. Ember CLI should strip these no-op function invocations from production builds by default.
+
+# Motivation
+
+Removing code that isn't required in production results in smaller and faster applications.
+
+# Detailed design
+
+The following framework function calls will be removed from ember-cli production builds by default:
+
+ * [`Ember.assert`](http://emberjs.com/api/#method_assert)
+ * [`Ember.debug`](http://emberjs.com/api/#method_debug)
+ * [`Ember.deprecate`](http://emberjs.com/api/#method_deprecate)
+ * [`Ember.info`](http://emberjs.com/api/#method_info)
+ * [`Ember.runInDebug`](http://emberjs.com/api/#method_runInDebug)
+ * [`Ember.warn`](http://emberjs.com/api/#method_warn)
+
+The API documentation will be updated where necessary to indicate that these function calls will be stripped from production.
+
+A Babel plugin will execute the removal of these function calls. It will either be:
+
+ * [babel-plugin-filter-imports](https://github.com/ember-cli/babel-plugin-filter-imports) with added support for defining module functions for removal ([spike here](https://github.com/GavinJoyce/babel-plugin-filter-imports/pull/1))
+ * a new babel plugin which works similarly to `babel-plugin-filter-imports`
+
+The babel plugin will affect the code of the current app or addon only and won't affect code in child or grandchild addons. As this change becomes part of the default ember-cli configuration, addons will adopt the code stripping as they upgrade to newer ember-cli versions.
+
+# How We Teach This
+
+This change doesn't bring any new functionality. Other than updating the Ember API docs, we don't need to make guide or other documentation changes. At the time of releasing, we may want to point out the possible side effects in a release blog post (see the _Drawbacks_ section below).
+
+If we want to expose the configuration options so that application authors can customize the settings, we can include a new section in the Ember CLI docs.
+
+# Drawbacks
+
+This may introduce an unexpected change in production builds as arguments that have side effects will no longer be executed. For example:
+
+```js
+Ember.assert('Some assertion', someSideEffect());
+```
+
+Currently, the `someSideEffect` function will be executed in production. When this RFC lands, it won't.
+
+# Alternatives
+
+An Ember addon could provide opt-in function stripping for applications that want it. If this RFC isn't deemed a good default for Ember CLI, that option should be explored.
+
+# Unresolved questions
+
+ * [ ] Should we allow the stripping configuration to be overridden in `ember-cli-build.js`? If so, what should the API be?
+ * [ ] Related to above, should we allow the feature to be disabled?
```

Comment:
Making this possible makes for annoying partial config in ember-cli-build.js (which things to strip) and in environment.js (which environments to strip in) and will be a source of numerous user-error bugs.

Since we will ship with defaults of some sort we do need to provide a way to modify/clobber those, and whatever that method is will be the way to turn it off (or just remove the addon). (Keeps the API cleaner.)